url882f91ec97_WenboCao_Microsoft_Drone_tgz-pJ8-org_ejml_alg_dense_decomposition_qr_StabilityQRDecompositionJ8/src/org/ejml/alg/dense/mult/TestMatrixMatrixMult.java

```diff
@@ -0,0 +1,215 @@
+                    DenseMatrix64F c_orig = RandomMatrices.createRandom(i,k, rand);
+
+                    DenseMatrix64F r_orig = RandomMatrices.createRandom(i,k,rand);
+
+                    MatrixMatrixMult.mult_small(a_orig,b_orig,r_orig);
+
+                    checkResults(a_orig,b_orig,c_orig,r_orig);
+                }
+            }
+        }
+    }
+
+    /**
+     * Sees if all the matrix multiplications produce the expected results against the provided
+     * known solution.
+     */
+    private void checkResults( DenseMatrix64F a_orig ,
+                               DenseMatrix64F b_orig ,
+                               DenseMatrix64F c_orig ,
+                               DenseMatrix64F r_orig )
+            throws InvocationTargetException, IllegalAccessException
+    {
+        double alpha = 2.5;
+
+        int numChecked = 0;
+        Method methods[] = MatrixMatrixMult.class.getMethods();
+
+        for( Method method : methods ) {
+            String name = method.getName();
+//            System.out.println(name);
+
+            // only look at function which perform matrix multiplications
+            if( !name.contains("mult") )
+                continue;
+
+            DenseMatrix64F a = a_orig.copy();
+            DenseMatrix64F b = b_orig.copy();
+            DenseMatrix64F c = c_orig.copy();
+
+            boolean add = name.contains("multAdd");
+            boolean hasAlpha = method.getGenericParameterTypes()[0] == double.class;
+
+            if( name.contains("TransAB")) {
+                transpose(a);
+                transpose(b);
+            } else if( name.contains("TransA")) {
+                transpose(a);
+            } else if( name.contains("TransB")) {
+                transpose(b);
+            }
+
+            DenseMatrix64F expected = r_orig.copy();
+            double []expectedData = expected.data;
+
+            if( hasAlpha ) {
+                for( int i = 0; i < expectedData.length; i++ ) {
+                    expectedData[i] *= alpha;
+                }
+            }
+            if( add ) {
+                for( int i = 0; i < expectedData.length; i++ ) {
+                    expectedData[i] += c_orig.get(i);
+                }
+            }
+
+            invoke(method,alpha,a,b,c);
+
+            EjmlUnitTests.assertEquals(expected,c,1e-12);
+            numChecked++;
+        }
+
+        assertEquals(numChecked,32);
+    }
+
+    private void transpose( DenseMatrix64F a ) {
+        DenseMatrix64F b = new DenseMatrix64F(a.numCols,a.numRows);
+        CommonOps.transpose(a,b);
+        a.setReshape(b);
+    }
+
+    public static void invoke(Method func,
+                              double alpha,
+                              DenseMatrix64F a, DenseMatrix64F b, DenseMatrix64F c)
+            throws IllegalAccessException, InvocationTargetException {
+        if( func.getParameterTypes().length == 3 ) {
+            func.invoke(null, a, b, c);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Catch `InvocationTargetException` explicitly when you call `Method.invoke()`. Use `Throwable.getCause()` or `getTargetException()` in the catch handler to access more information about the underlying exception. See the following diffs on GitHub that make this code change: [diff1](https://github.com/EsfingeFramework/gamification/commit/20792047e6c69457cca03e877f2ada578188f32a#diff-6808c12fe9ddaea042db84f95dd5ba78adab99d0cb18bb4d85916a8723530768L21), [diff2](https://github.com/cucumber-attic/cuke4duke/commit/f4821d400cf2590126f1298a8dff9f7a33d45dc5#diff-80adaffc20c0e6e7d2e42312386ffe1ddc10213bf5b8e8919895fd85d666c8deL34)

Learn more: [InvocationTargetException in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/InvocationTargetException.html)